TypeCobol/Compiler/CodeElements/Entry/DataType.cs

```diff
@@ -9,58 +9,79 @@
 
 namespace TypeCobol.Compiler.CodeElements
 {
-	public class DataType : ICobolLanguageLevel, IVisitable {
-		public string Name { get; private set; }
+    public class DataType : ICobolLanguageLevel, IVisitable, IEquatable<DataType>
+    {
+        public string Name { get; }
         public RestrictionLevel RestrictionLevel { get; internal set; }
-		public CobolLanguageLevel CobolLanguageLevel  { get; private set; }
+        public CobolLanguageLevel CobolLanguageLevel  { get; }
 
-		public DataType([NotNull] string name, RestrictionLevel restrictionLevel = RestrictionLevel.WEAK, CobolLanguageLevel cobolLanguageLevel = CobolLanguageLevel.Cobol85) {
-			Name = name;
-			if (name == null) throw new ArgumentNullException();
-		    this.RestrictionLevel = restrictionLevel;
+        public DataType([NotNull] string name, RestrictionLevel restrictionLevel = RestrictionLevel.WEAK, CobolLanguageLevel cobolLanguageLevel = CobolLanguageLevel.Cobol85) {
+            Name = name;
+            if (name == null) throw new ArgumentNullException();
+            this.RestrictionLevel = restrictionLevel;
             this.CobolLanguageLevel = cobolLanguageLevel;
-		}
+        }
 
-		public override string ToString() { return Name; }
+        public override string ToString() { return Name; }
 
-		public override int GetHashCode() { return Name.GetHashCode(); }
-	    public bool AcceptASTVisitor(IASTVisitor astVisitor) {
-	        return astVisitor.Visit(this);
-	    }
+        public bool AcceptASTVisitor(IASTVisitor astVisitor) {
+            return astVisitor.Visit(this);
+        }
 
-	    public override bool Equals(object obj) {
-			var other = obj as DataType;
-			if (other == null) return false;
-			return other == this;
-		}
+        public override bool Equals(object obj) {
+            return Equals(this, obj as DataType);
```

Comment:
This implementation of Equals doesn't respects our guideline :
https://github.com/TypeCobolTeam/TypeCobol/issues/1539


Can you adapt the code,

Reply:
It's already done.
public override bool Equals(object obj) {
            return Equals(this, obj as DataType);
        }

        public bool Equals(DataType dataType)
        {
            return Equals(this, dataType);
        }

        private static bool Equals([CanBeNull]DataType x, [CanBeNull]DataType y)
        {
            if (Object.ReferenceEquals(x, y)) return true;
            if (Object.ReferenceEquals(x, null)) return false;
            if (Object.ReferenceEquals(null, y)) return false;

            return string.Equals(x.Name, y.Name, StringComparison.OrdinalIgnoreCase)
                   && x.RestrictionLevel.Equals(y.RestrictionLevel)
                   && x.CobolLanguageLevel.Equals(y.CobolLanguageLevel);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = 13;
                hashCode = (hashCode * 397) ^ Name.GetHashCode();
                hashCode = (hashCode * 397) ^ RestrictionLevel.GetHashCode();
                hashCode = (hashCode * 397) ^ CobolLanguageLevel.GetHashCode();

                return hashCode;
            }
        }